# Lab | Data Aggregation and Filtering

In this challenge, we will continue to work with customer data from an insurance company. We will use the dataset called marketing_customer_analysis.csv, which can be found at the following link:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis.csv

This dataset contains information such as customer demographics, policy details, vehicle information, and the customer's response to the last marketing campaign. Our goal is to explore and analyze this data by first performing data cleaning, formatting, and structuring.

1. Create a new DataFrame that only includes customers who have a total_claim_amount greater than $1,000 and have a response of "Yes" to the last marketing campaign.

In [3]:
import pandas as pd
url = 'https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis.csv'
df = pd.read_csv(url)
df.rename(columns= {'Total Claim Amount': 'Total_Claim_Amount'}, inplace=True)
df.rename(columns= {'Policy Type': 'Policy_Type'}, inplace=True)
df.rename(columns= {'Customer Lifetime Value': 'Customer_Lifetime_Value'}, inplace=True)

In [ ]:
df.drop(['Unnamed: 0', 'Customer', 'State', 'Customer Lifetime Value', 'Coverage', 'Education',
        'Effective To Date',
       'EmploymentStatus', 'Gender', 'Income', 'Location Code',
       'Marital Status', 'Monthly Premium Auto', 'Months Since Last Claim',
       'Months Since Policy Inception', 'Number of Open Complaints',
       'Number of Policies', 'Policy Type', 'Policy', 'Renew Offer Type',
       'Sales Channel', 'Vehicle Class', 'Vehicle Size',
       'Vehicle Type'], axis=1, inplace=True)


#grouped_response = df.groupby('Response')
#grouped_response.get_group('Yes').head()
filtered_df = df[(df['Total_Claim_Amount'] > 1000) & (df['Response'] == 'Yes')]
new_df = filtered_df
new_df

2. Using the original Dataframe, analyze the average total_claim_amount by each policy type and gender for customers who have responded "Yes" to the last marketing campaign. Write your conclusions.

In [ ]:
average_claim_by_policy_gender = df[df['Response'] == 'Yes'].groupby(['Policy', 'Gender'])['Total_Claim_Amount'].mean()
average_claim_by_policy_gender

Out of all associates whom answered our marketing campagain positively we can see that total amount of claims by female customers is almost 464k while male in same cases total is 8,5% lower,being 426k, which creates assumption that our female customers are not only active participants in our marketing solutions ,but also tend to get higher value claims.


3. Analyze the total number of customers who have policies in each state, and then filter the results to only include states where there are more than 500 customers.

In [ ]:
customer_count = df.groupby('State').size()
states500 = customer_count[customer_count > 500]
states500

4. Find the maximum, minimum, and median customer lifetime value by education level and gender. Write your conclusions.

In [5]:
summary_stats = df.groupby(['Education', 'Gender'])['Customer_Lifetime_Value'].agg(['max', 'min', 'median'])
summary_stats

max          min       median
Education            Gender                                       
Bachelor             F       73225.95652  1904.000852  5640.505303
                     M       67907.27050  1898.007675  5548.031892
College              F       61850.18803  1898.683686  5623.611187
                     M       61134.68307  1918.119700  6005.847375
Doctor               F       44856.11397  2395.570000  5332.462694
                     M       32677.34284  2267.604038  5577.669457
High School or Below F       55277.44589  2144.921535  6039.553187
                     M       83325.38119  1940.981221  6286.731006
Master               F       51016.06704  2417.777032  5729.855012
                     M       50568.25912  2272.307310  5579.099207

Standart CLV for male with High School or bellow education on average is 6286, but the CLV deviates from 1940 to 83325 (190.895% differenc) ,this wide range indicates other factors having a big effect to the results, meanwhile the biggest range in females were noticed mostly in College and Bachelor (189.863% difference). In general higher educations like - Master's and Doctor has very little difference in its range and/or gender.

## Bonus

5. The marketing team wants to analyze the number of policies sold by state and month. Present the data in a table where the months are arranged as columns and the states are arranged as rows.

6.  Display a new DataFrame that contains the number of policies sold by month, by state, for the top 3 states with the highest number of policies sold.

*Hint:*
- *To accomplish this, you will first need to group the data by state and month, then count the number of policies sold for each group. Afterwards, you will need to sort the data by the count of policies sold in descending order.*
- *Next, you will select the top 3 states with the highest number of policies sold.*
- *Finally, you will create a new DataFrame that contains the number of policies sold by month for each of the top 3 states.*

7. The marketing team wants to analyze the effect of different marketing channels on the customer response rate.

Hint: You can use melt to unpivot the data and create a table that shows the customer response rate (those who responded "Yes") by marketing channel.

External Resources for Data Filtering: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9

In [ ]:
# your code goes here